### PR TITLE
Persist offer history

### PR DESCRIPTION
**Pull request by TheCharlatan**
Adds persistance for offers. Offers can now be written to a separate table in the lmdb database.

They can be retrieved by selecting their status. The `OfferStatus` and `OfferStatusSelector` are separated to allow for extension of their criteria. For example saving addresses in the `Open` status, or `SwapId`s in the `InProgress` status, and selecting offers based on other common criteria, like amounts.

### CODE REVIEW

File: src/rpc/request.rs [hunk 1]
```diff
@@ -704,9 +704,66 @@ pub enum Request {
...
+)]
+pub enum OfferStatus {
+    #[display("Open")]
+    Open,
+    #[display("In Progress")]
+    InProgress,
+    #[display("Ended({0})")]
+    Ended(Outcome),
+}
+
```
Comment (zkao) [inline src/rpc/request.rs:760]: `Expired`?

**Review by zkao — APPROVED** [review]
lgtm, minor comments only

File: src/databased/runtime.rs [hunk 2]
```diff
@@ -466,18 +483,71 @@ struct Database(lmdb::Environment);
 
 const LMDB_CHECKPOINTS: &str = "checkpoints";
 const LMDB_ADDRESSES: &str = "addresses";
+const LMDB_OFFER_HISTORY: &str = "history";
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/databased/runtime.rs:486]: Ah, good point.

File: src/rpc/request.rs (hunk 1 above)
Comment (TheCharlatan) [reply src/rpc/request.rs:760]: `Expired` has its own semantics for me, I'd rather think of it as a variant of the `Outcome` enum.

File: src/databased/runtime.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/databased/runtime.rs]: Done in 02aee9184fa675c5d9258858e005d41cb8d4377c

File: src/rpc/request.rs (hunk 1 above)
Comment (zkao) [reply src/rpc/request.rs:774]: I see. For me `Outcome` is about a swap, not an offer, and an offer that was never taken never becomes an `Outcome`, its just an expired offer

Comment (zkao) [reply src/rpc/request.rs:774]: For me a `Trade` is a swap, not an `Offer`, and only a trade has an `Outcome`, so I don't see how that helps

Comment (zkao) [reply src/rpc/request.rs:774]: But it doesnt matter for now anyway

Comment (h4sh3d) [reply src/rpc/request.rs:774]: `Trade` would be different than `Offer`: this should allow use to have a proper `Offer` type (with room for negotiation) and a `Trade` type (close to what we currently have in `Offer`) that represent all the parameters of a swap.

`Offer` is kind of in between currently.

**Pull request merged**